### ISSUE TITLE
This Realm instance has already been closed - getting this as soon as the fragment is destroyed.

### ISSUE THREAD

**Issue by anjanibansal** [realm/realm-java#2807]
Hi,  
I'm using a 'RealmBaseAdapter' to populate a 'ListView' in a fragment. The activity that hosts the fragment does not have a Realm instance.  
I create the Realm instance in 'onStart()' and close it in 'onStop()' of the fragment.  
I just don't know where am I going wrong but as soon as I move out of that fragment, like replace it with other fragment (which has no Realm instance) or just close the host activity, I'm getting the above exception.  
Here's my log:  
  
```
FATAL EXCEPTION: main
Process: com.anjani.myapp, PID: 31959
java.lang.IllegalStateException: This Realm instance has already been closed, making it unusable.
at io.realm.BaseRealm.checkIfValid(BaseRealm.java:344)
at io.realm.RealmResults.isLoaded(RealmResults.java:748)
at io.realm.RealmResults.size(RealmResults.java:379)
at io.realm.RealmBaseAdapter.getCount(RealmBaseAdapter.java:70)
at com.nhaarman.listviewanimations.BaseAdapterDecorator.getCount(BaseAdapterDecorator.java:109)
at android.widget.AdapterView.isScrollableForAccessibility(AdapterView.java:1009)
at android.widget.AdapterView.onInitializeAccessibilityNodeInfoInternal(AdapterView.java:984)
at android.widget.AbsListView.onInitializeAccessibilityNodeInfoInternal(AbsListView.java:1496)
at android.widget.ListView.onInitializeAccessibilityNodeInfoInternal(ListView.java:3909)
at android.view.View.onInitializeAccessibilityNodeInfo(View.java:6084)
at android.view.View.createAccessibilityNodeInfoInternal(View.java:6043)
at android.view.View.createAccessibilityNodeInfo(View.java:6028)
at android.view.accessibility.AccessibilityRecord.setSource(AccessibilityRecord.java:145)
at android.view.accessibility.AccessibilityRecord.setSource(AccessibilityRecord.java:119)
at android.view.View.onInitializeAccessibilityEventInternal(View.java:5980)
at android.widget.AdapterView.onInitializeAccessibilityEventInternal(AdapterView.java:994)
at android.view.View.onInitializeAccessibilityEvent(View.java:5968)
at android.view.View.sendAccessibilityEventUncheckedInternal(View.java:5833)
at android.view.View.sendAccessibilityEventUnchecked(View.java:5818)
at android.view.View$SendViewStateChangedAccessibilityEvent.run(View.java:22324)
at android.view.View$SendViewStateChangedAccessibilityEvent.runOrPost(View.java:22357)
at android.view.View.notifyViewAccessibilityStateChangedIfNeeded(View.java:8784)
at android.view.View.onFocusChanged(View.java:5701)
at android.widget.AbsListView.onFocusChanged(AbsListView.java:1959)
at android.widget.ListView.onFocusChanged(ListView.java:3580)
at android.view.View.clearFocusInternal(View.java:5605)
at android.view.View.unFocus(View.java:5638)
at android.view.ViewGroup.unFocus(ViewGroup.java:994)
at android.view.ViewGroup.unFocus(ViewGroup.java:996)
at android.view.ViewGroup.unFocus(ViewGroup.java:996)
at android.view.ViewGroup.removeViewInternal(ViewGroup.java:4590)
at android.view.ViewGroup.removeViewInternal(ViewGroup.java:4576)
at android.view.ViewGroup.removeView(ViewGroup.java:4507)
at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1068)
at android.app.FragmentManagerImpl.removeFragment(FragmentManager.java:1253)
at android.app.BackStackRecord.run(BackStackRecord.java:744)
at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1535)
at android.app.FragmentManagerImpl$1.run(FragmentManager.java:482)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:148)
at android.app.ActivityThread.main(ActivityThread.java:5417)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

```  
Here's some code:  
  
```
@Override
    public void onStart() {
        super.onStart();
        mRealm = Realm.getDefaultInstance();
        list = mRealm.where(Data.class).findAll();
        list.sort("name", Sort.ASCENDING);

        mAdapter = new MyAdapter(mContext, list, true);
        animationAdapter = new AlphaInAnimationAdapter(mAdapter);
        animationAdapter.setAbsListView(getListView());

        getListView().setAdapter(animationAdapter);
}  
  
@Override
    public void onStop() {
        super.onStop();
        mRealm.close();
    }  
  
 @Override
    public void onDestroyView() {
        ButterKnife.unbind(this);
        super.onDestroyView();
    }  

```
The fragment is showing up perfectly fine. The error comes when I move out of it.  
I'm using version 0.87.5 on Android Studio 2.0. Also, I'm using third party libraries- Butterknife and ListViewAnimations in the fragment. I've tried closing the Realm in 'onDestroyView()' also, and the same thing happens.  
  
Please guide me on what am I doing wrong.

**Comment by stk1m1** [comment]
Hi @anjanibansal,

I've noticed that `mRealm` instance gets closed on [`onStop()`](http://developer.android.com/reference/android/app/Activity.html#onStop()), and can also see you paired up `onStart()` and `onStop()` together for controling the lifecycle of `mRealm`, but I'd suggest to move the control to `onActivityCreated()` and `onDestroy()` instead.

`Fragment.onStop()` is generally controlled by `Activity.onClose()`, and according to Android Activity Doc, `Activity.onClose()` is

> Called when you are no longer visible to the user. You will next receive either onRestart(), onDestroy(), or nothing, depending on later user activity.

I cannot see how your Fragment behaves in its parent Activity, but my hypothesis is that `ListView` in the Fragment is trying to achieve something (`getCount()`?) whileas the parent Activity already closes itself as it gets hidden. 

We also provide handcrafted, carefully assorted examples [here](https://github.com/realm/realm-java/tree/master/examples/threadExample). Hope it helps.

**Comment by anjanibansal** [comment]
Hey,  
Thanks for the response. I did try moving the control to `onActivityCreated()` and `onDestroy()` but that didn't solve the issue.  
The issue is now resolved though. I added the following line just before closing the Realm instance.  
  
        getListView().setAdapter(null);
  
This solved the issue. I guess one of the third party library was trying to access the adapter after the Realm instance got closed.  
  
Anyway, thanks for responding.
